File: lib/rgfa/field.rb
require "json"
require_relative "alignment"
require_relative "byte_array"
require_relative "error"
require_relative "field_array"
require_relative "line"
require_relative "numeric_array"

RGFA::Field = Module.new

require_relative "field/alignment_gfa1.rb"
require_relative "field/alignment_gfa2.rb"
require_relative "field/alignment_list_gfa1.rb"
require_relative "field/byte_array.rb"
require_relative "field/char.rb"
require_relative "field/comment.rb"
require_relative "field/custom_record_type.rb"
require_relative "field/float.rb"
require_relative "field/generic.rb"
require_relative "field/identifier_gfa2.rb"
require_relative "field/identifier_list_gfa2.rb"
require_relative "field/oriented_identifier_list_gfa1.rb"
require_relative "field/oriented_identifier_list_gfa2.rb"
require_relative "field/integer.rb"
require_relative "field/json.rb"
require_relative "field/numeric_array.rb"
require_relative "field/optional_identifier_gfa2.rb"
require_relative "field/optional_integer.rb"
require_relative "field/orientation.rb"
require_relative "field/oriented_identifier_gfa2.rb"
require_relative "field/path_name_gfa1.rb"
require_relative "field/position_gfa1.rb"
require_relative "field/position_gfa2.rb"
require_relative "field/segment_name_gfa1.rb"
require_relative "field/sequence_gfa1.rb"
require_relative "field/sequence_gfa2.rb"
require_relative "field/string.rb"

# Decoding, validation and encoding of GFA fields.
#
# For each datatype a module under field/ exists, which defines
# the following methods as module functions:
#
# unsafe_decode => parses an ASSUMED VALID string representation to
#                  an appropriate Ruby object
#                  - faster or as fast as decode()
#                  - if the assumption is not met, sometimes it will
#                  raise an exception, sometimes it will return an
#                  invalid object
#
# decode => parses a string representation to an appropriate Ruby object
#           - if the string is invalid, an exception is raised
#           - the returned object is guaranteed to be valid
#
# validate_encoded => validates a string representation
#                     - raises RGFA::FormatError if invalid
#
# validate_decoded => validates a non-string field content
#                     - raises exception if its state is invalid
#
# unsafe_encode => encodes an ASSUMED VALID field to the string representation;
#                  - faster or as fast as encode()
#                  - if the assumption is not met, sometimes it will
#                  raise an exception, sometimes it will return an
#                  invalid string representation
#
# encode => encodes a field to its string representation;
#           - raises an exception if the field content is invalid
#           - the string representation is guaranteed to be valid;
#
# Everything in the RGFA::Field module is API private. The user will not call
# these methods directly, and use instead methods of RGFA::Line.
# Also: code in line.rb should not call the functions of the submodules
# defined in the field/* files directly, but rather call the functions of
# the submodules defined in this file, ie RGFA::Field::Parser,
# RGFA::Field::Validator, RGFA::Field::Writer.
#
# @api private
module RGFA::Field

  # Symbol representing a GFA1-specific datatype for positional fields
  GFA1_POSFIELD_DATATYPE = [
                             :alignment_gfa1,
                             :alignment_list_gfa1,
                             :oriented_identifier_list_gfa1,
                             :position_gfa1,
                             :segment_name_gfa1,
                             :sequence_gfa1,
                             :path_name_gfa1,
                           ]

  # Symbol representing a GFA2-specific datatype for positional fields
  GFA2_POSFIELD_DATATYPE = [
                             :alignment_gfa2,
                             :generic,
                             :identifier_gfa2,
                             :oriented_identifier_gfa2,
                             :identifier_list_gfa2,
                             :oriented_identifier_list_gfa2,
                             :optional_identifier_gfa2,
                             :position_gfa2,
                             :custom_record_type,
                             :sequence_gfa2,
                             :optional_integer,
                           ]

  # Symbol representing a datatype for positional fields common to GFA1 and GFA2
  GFAX_POSFIELD_DATATYPE = [:comment, :orientation]

  # Symbol representing a datatype for positional fields
  POSFIELD_DATATYPE = GFA1_POSFIELD_DATATYPE +
                      GFA2_POSFIELD_DATATYPE +
                      GFAX_POSFIELD_DATATYPE

  # A symbol representing a datatype for tags
  TAG_DATATYPE = [:A, :i, :f, :Z, :J, :H, :B]

  # A symbol representing a valid datatype
  FIELD_DATATYPE = TAG_DATATYPE + POSFIELD_DATATYPE

  FIELD_MODULE = {
    :alignment_gfa1                => RGFA::Field::AlignmentGFA1,
    :alignment_gfa2                => RGFA::Field::AlignmentGFA2,
    :alignment_list_gfa1           => RGFA::Field::AlignmentListGFA1,
    :comment                       => RGFA::Field::Comment,
    :custom_record_type            => RGFA::Field::CustomRecordType,
    :generic                       => RGFA::Field::Generic,
    :identifier_gfa2               => RGFA::Field::IdentifierGFA2,
    :identifier_list_gfa2          => RGFA::Field::IdentifierListGFA2,
    :oriented_identifier_list_gfa1 => RGFA::Field::OrientedIdentifierListGFA1,
    :oriented_identifier_list_gfa2 => RGFA::Field::OrientedIdentifierListGFA2,
    :optional_identifier_gfa2      => RGFA::Field::OptionalIdentifierGFA2,
    :oriented_identifier_gfa2      => RGFA::Field::OrientedIdentifierGFA2,
    :optional_integer              => RGFA::Field::OptionalInteger,
    :orientation                   => RGFA::Field::Orientation,
    :path_name_gfa1                => RGFA::Field::PathNameGFA1,
    :position_gfa1                 => RGFA::Field::PositionGFA1,
    :position_gfa2                 => RGFA::Field::PositionGFA2,
    :segment_name_gfa1             => RGFA::Field::SegmentNameGFA1,
    :sequence_gfa1                 => RGFA::Field::SequenceGFA1,
    :sequence_gfa2                 => RGFA::Field::SequenceGFA2,
    :H    => RGFA::Field::ByteArray,
    :A    => RGFA::Field::Char,
    :f    => RGFA::Field::Float,
    :i    => RGFA::Field::Integer,
    :J    => RGFA::Field::JSON,
    :B    => RGFA::Field::NumericArray,
    :Z    => RGFA::Field::String,
  }

  # Encoding of Ruby objects to GFA string representation
  # @api private
  module Writer

    # Encode a Ruby object into a GFA field. The ruby object can be
    # either an encoded GFA field (in which case, at most it is validated,
    # see +safe+, but not encoded) or an object of a class compatible
    # with the specified datatype, if a datatype is specified (see +datatype+),
    # e.g. Integer # for i fields.
    # @param datatype [RGFA::Field::FIELD_DATATYPE] datatype to use. If no
    #   datatype is specified, any class will do and the default datatype
    #   will be chosen (see RGFA::DefaultDatatype module).
    # @param fieldname [String] fieldname, for error messages (optional)
    # @param safe [Boolean] <i>(defaults to: +true+)</i> if +true+, the safe
    #   version of the encode function is used, which guarantees that the
    #   resulting data is valid; if +false+, the unsafe version is used,
    #   which, for some datatypes, skips validations in order to be faster
    #   than the safe version
    # @raise [RGFA::TypeError] if an unknown datatype is specified
    # @raise [RGFA::ValueError] if the object value is invalid for the datatype
    # @raise [RGFA::FormatError] if the object syntax is invalid for the
    #   datatype (eg for invalid encoded strings, if +safe+ is set)
    # @raise [RGFA::TypeError] if the type of the object and the datatype
    #   are not compatible
    def to_gfa_field(datatype: nil, safe: true, fieldname: nil)
      datatype ||= default_gfa_tag_datatype
      mod = RGFA::Field::FIELD_MODULE[datatype]
      if mod.nil?
        fieldnamemsg = fieldname ? "Field: #{fieldname}\n" : ""
        contentmsg = "Content: #{self.inspect}\n"
        raise RGFA::TypeError,
          fieldnamemsg +
          contentmsg +
          "Datatype unknown: #{datatype.inspect}"
      end
      begin
        if safe
          mod.encode(self)
        else
          mod.unsafe_encode(self)
        end
      rescue => err
        fieldnamemsg = fieldname ? "Field: #{fieldname}\n" : ""
        contentmsg = "Content: #{self.inspect}\n"
        datatypemsg = "Datatype: #{datatype}\n"
        raise err.class,
              fieldnamemsg +
              datatypemsg +
              contentmsg +
              err.message
      end
    end

    # Representation of the data as a GFA tag +xx:d:content+, where +xx+ is
    # the tag name and +d+ is the datatype.
    # @param fieldname [Symbol] the tag name
    # @param datatype [RGFA::Field::TAG_DATATYPE] (<i>defaults to: the value
    #  returned by {#default_gfa_tag_datatype}</i>)
    def to_gfa_tag(fieldname, datatype: default_gfa_tag_datatype)
      return "#{fieldname}:#{datatype}:"+
      "#{to_gfa_field(datatype: datatype, fieldname: fieldname)}"
    end
  end

  # Decoding of the GFA string representations into Ruby objects
  # @api private
  module Parser

    # Parse a GFA string representation and decodes it into a Ruby object
    # @param datatype [RGFA::Field::FIELD_DATATYPE] the datatype to use
    # @param safe [Boolean] <i>(defaults to: +true+)</i> if +true+ the safe
    #   version of the decode function for the datatype is used, which
    #   validates the content of the string; if +false+, the string is
    #   assumed to be valid and decoded into a value accordingly, which may
    #   result in invalid values (but may be faster than the safe decoding)
    # @param fieldname [String] fieldname, for error messages (optional)
    # @param line [#to_s] line content, for error messages (optional)
    # @raise [RGFA::TypeError] if the specified datatype is unknown
    # @raise [RGFA::FormatError] if the string syntax is not valid
    # @raise [RGFA::ValueError] if the decoded value is not valid
    def parse_gfa_field(datatype,
                        safe: true,
                        fieldname: nil,
                        line: nil)
      mod = RGFA::Field::FIELD_MODULE[datatype]
      if mod.nil?
        begin
          linemsg = line ? "Line content: #{line.to_s}\n" : ""
        rescue
          linemsg = ""
        end
        fieldnamemsg = fieldname ? "Field: #{fieldname}\n" : ""
        contentmsg = "Content: #{self}\n"
        raise RGFA::TypeError,
          linemsg +
          fieldnamemsg +
          contentmsg +
          "Datatype unknown: #{datatype.inspect}"
      end
      begin
        if safe
          mod.decode(self)
        else
          mod.unsafe_decode(self)
        end
      rescue => err
        begin
          linemsg = line ? "Line content: #{line.to_s}\n" : ""
        rescue
          linemsg = ""
        end
        fieldnamemsg = fieldname ? "Field: #{fieldname}\n" : ""
        contentmsg = "Content: #{self}\n"
        datatypemsg = "Datatype: #{datatype}\n"
        raise err.class,
              linemsg +
              fieldnamemsg +
              datatypemsg +
              contentmsg +
              err.message
      end
    end

    # Parses a GFA tag in the form +xx:d:content+ into its components.
    # The +content+ is not decoded (see #parse_gfa_field).
    # @raise [RGFA::FormatError] if the string does not represent
    #   a valid GFA tag
    # @return [Array(Symbol, RGFA::Field::FIELD_DATATYPE, String)]
    #   the parsed content of the field
    def parse_gfa_tag
      if self =~ /^([A-Za-z][A-Za-z0-9]):([AifZJHB]):(.+)$/
        return $1.to_sym, $2.to_sym, $3
      else
        raise RGFA::FormatError,
          "Expected GFA tag, found: #{self.inspect}"
      end
    end

  end

  # Validates the content of a GFA field, which can be a GFA string
  # representation or a Ruby object, according to the field datatype.
  # @api private
  module Validator

    # Validates a GFA string representation according to the field datatype.
    # @!macro [new] validate_gfa_field
    #   @raise [RGFA::TypeError] if an unknown datatype is specified
    #   @param datatype [RGFA::Field::FIELD_DATATYPE] the datatype to use
    #   @param fieldname [String] fieldname, for error messages (optional)
    #   @raise [RGFA::FormatError] if the object type or content
    #     is not compatible to the provided datatype
    #   @return [void]
    # @api private
    module Encoded
      def validate_gfa_field(datatype, fieldname=nil)
        mod = RGFA::Field::FIELD_MODULE[datatype]
        if mod.nil?
          raise RGFA::TypeError,
            "Datatype unknown: #{datatype.inspect}"
        end
        mod::validate_encoded(self)
      end
    end

    # Validates a non-string Ruby object field content
    # according to the field datatype.
    # @!macro validate_gfa_field
    # @api private
    module Decoded
      def validate_gfa_field(datatype, fieldname=nil)
        mod = RGFA::Field::FIELD_MODULE[datatype]
        if mod.nil?
          raise RGFA::TypeError,
            "Datatype unknown: #{datatype.inspect}"
        end
        mod::validate_decoded(self)
      end
    end

  end

end

class Object
  include RGFA::Field::Writer
  include RGFA::Field::Validator::Decoded
end

class String
  include RGFA::Field::Parser
  include RGFA::Field::Validator::Encoded
end

#
# This module specifies default datatypes for GFA tags
# for the core classes.
#
# Custom classes shall define the default_gfa_tag_datatype
# function in their class definition and not here.
#
# @api private
module RGFA::DefaultDatatypes

  module Object
    # @!macro [new] gfa_datatype
    #   GFA tag datatype to use, if none is provided
    #   @return [RGFA::Field::TAG_DATATYPE]
    def default_gfa_tag_datatype; :Z; end
  end

  module Fixnum
    # @!macro gfa_datatype
    def default_gfa_tag_datatype; :i; end
  end

  module Float
    # @!macro gfa_datatype
    def default_gfa_tag_datatype; :f; end
  end

  module Hash
    # @!macro gfa_datatype
    def default_gfa_tag_datatype; :J; end
  end

  module Array
    # @!macro gfa_datatype
    def default_gfa_tag_datatype
      (all?{|i|i.kind_of?(Integer)} or all?{|i|i.kind_of?(Float)}) ? :B : :J
    end
  end

end

class Object; include RGFA::DefaultDatatypes::Object; end
class Fixnum; include RGFA::DefaultDatatypes::Fixnum; end
class Float;  include RGFA::DefaultDatatypes::Float;  end
class Hash;   include RGFA::DefaultDatatypes::Hash;   end
class Array;  include RGFA::DefaultDatatypes::Array;  end

File: gfapy/field/field.py
import gfapy
import builtins
from .validator import Validator
from .parser import Parser
from .writer import Writer
from . import alignment_gfa1                as field_alignment_gfa1
from . import alignment_gfa2                as field_alignment_gfa2
from . import alignment_list_gfa1           as field_alignment_list_gfa1
from . import byte_array                    as field_byte_array
from . import char                          as field_char
from . import comment                       as field_comment
from . import custom_record_type            as field_custom_record_type
from . import float                         as field_float
from . import generic                       as field_generic
from . import identifier_gfa2               as field_identifier_gfa2
from . import oriented_identifier_gfa2      as field_oriented_identifier_gfa2
from . import identifier_list_gfa2          as field_identifier_list_gfa2
from . import integer                       as field_integer
from . import json                          as field_json
from . import numeric_array                 as field_numeric_array
from . import optional_identifier_gfa2      as field_optional_identifier_gfa2
from . import optional_integer              as field_optional_integer
from . import orientation                   as field_orientation
from . import oriented_identifier_list_gfa1 as field_oriented_identifier_list_gfa1
from . import oriented_identifier_list_gfa2 as field_oriented_identifier_list_gfa2
from . import path_name_gfa1                as field_path_name_gfa1
from . import position_gfa1                 as field_position_gfa1
from . import position_gfa2                 as field_position_gfa2
from . import segment_name_gfa1             as field_segment_name_gfa1
from . import sequence_gfa1                 as field_sequence_gfa1
from . import sequence_gfa2                 as field_sequence_gfa2
from . import string                        as field_string

class Field(Validator, Parser, Writer):
  """
  Support for the decoding, validation and encoding of data in GFA fields.

  Classes are defined (and imported here) for each type of field (positional and
  tags) defined in the GFA specifications. The field definition classes
  implement the following methods:

  * ``decode(str)``: decodes the content of a GFA field into a Python object
      which represents the value; the content of the field is validated and
      the returned object is guaranteed to be valid
  * ``unsafe_decode(str)``: an optional method, which decodes the content
      of the string faster than decode(), but does not perform validations
  * ``validate_encoded(str)``: validates the content of a GFA field, when this
      is in its string form; it can be called by the decode() method
  * ``validate_decoded(obj)``: validates the content of a GFA field, when this
      is in a non-string form; it can be called by the encode() method
  * ``encode(obj)``: takes a non-string content of a GFA field and converts it
      in its string representation according to the GFA specification; the
      returned string is guaranteed to be valid
  * ``unsafe_encode(obj)``: an optional method, which encodes the content
      of a non-string field faster than encode(), but does not perform
      validations

  Notes:
    The library user does not call these methods directly, as the interaction
    is done using the interface of the `~gfapy.line.line.Line` class.
    However, an user may define classes for custom datatypes, to be used with
    custom record types.
  """

  _default_tag_datatypes = [
    (builtins.int , "i"),
    (builtins.float , "f"),
    (builtins.dict , "J"),
    (builtins.list , "J"),
    (builtins.object , "Z")
  ]
  """Default tag datatype to be used if the value is of a built-in class.

  For non build-in classes, the _default_gfa_tag_datatype() method of the
  class is called instead.
  """

  GFA1_POSFIELD_DATATYPE = [
                             "alignment_gfa1",
                             "alignment_list_gfa1",
                             "oriented_identifier_list_gfa1",
                             "position_gfa1",
                             "segment_name_gfa1",
                             "sequence_gfa1",
                             "path_name_gfa1",
                           ]
  """The names of the GFA1-specific datatypes for positional fields."""

  GFA2_POSFIELD_DATATYPE = [
                             "alignment_gfa2",
                             "generic",
                             "identifier_gfa2",
                             "oriented_identifier_gfa2",
                             "identifier_list_gfa2",
                             "oriented_identifier_list_gfa2",
                             "optional_identifier_gfa2",
                             "position_gfa2",
                             "custom_record_type",
                             "sequence_gfa2",
                             "optional_integer",
                           ]
  """The names of the GFA2-specific datatypes for positional fields."""

  GFAX_POSFIELD_DATATYPE = [ "comment", "orientation" ]
  """The names of the non version-specific datatypes for positional fields."""

  POSFIELD_DATATYPE = GFA1_POSFIELD_DATATYPE + \
                      GFA2_POSFIELD_DATATYPE + \
                      GFAX_POSFIELD_DATATYPE
  """The names of all datatypes for positional fields."""

  TAG_DATATYPE = ["A", "i", "f", "Z", "J", "H", "B"]
  """The names of all datatypes for tags."""

  FIELD_DATATYPE = TAG_DATATYPE + POSFIELD_DATATYPE
  """The names of all datatypes for positional fields and tags."""

  FIELD_MODULE = {
    "alignment_gfa1"                 : field_alignment_gfa1,
    "alignment_gfa2"                 : field_alignment_gfa2,
    "alignment_list_gfa1"            : field_alignment_list_gfa1,
    "byte_array"                     : field_byte_array,
    "char"                           : field_char,
    "comment"                        : field_comment,
    "custom_record_type"             : field_custom_record_type,
    "float"                          : field_float,
    "generic"                        : field_generic,
    "identifier_gfa2"                : field_identifier_gfa2,
    "oriented_identifier_gfa2"       : field_oriented_identifier_gfa2,
    "identifier_list_gfa2"           : field_identifier_list_gfa2,
    "integer"                        : field_integer,
    "json"                           : field_json,
    "numeric_array"                  : field_numeric_array,
    "optional_identifier_gfa2"       : field_optional_identifier_gfa2,
    "optional_integer"               : field_optional_integer,
    "orientation"                    : field_orientation,
    "oriented_identifier_list_gfa1"  : field_oriented_identifier_list_gfa1,
    "oriented_identifier_list_gfa2"  : field_oriented_identifier_list_gfa2,
    "path_name_gfa1"                 : field_path_name_gfa1,
    "position_gfa1"                  : field_position_gfa1,
    "position_gfa2"                  : field_position_gfa2,
    "segment_name_gfa1"              : field_segment_name_gfa1,
    "sequence_gfa1"                  : field_sequence_gfa1,
    "sequence_gfa2"                  : field_sequence_gfa2,
    "string"                         : field_string,
    "H"    : field_byte_array,
    "A"    : field_char,
    "f"    : field_float,
    "i"    : field_integer,
    "J"    : field_json,
    "B"    : field_numeric_array,
    "Z"    : field_string,
  }
  """Assignment of a class for the parsing, validation and encoding of data.

  The dictionary contains keys for each GFA datatype; the value is a class name,
  which provides the encoding, decoding and validation methods.

  For simplicity of use, tag datatypes are present twice, once with a
  one-letter symbol (such as i) and once with a longer labe; (such as integer).
  """

  # Returns the default GFA tag for the given object.
  @staticmethod
  def _get_default_gfa_tag_datatype(obj):
    """Default GFA tag datatype for a given object

    Parameters:
      obj : an object of any Python class

    Returns:
      str : the identifier of a datatype (one of the keys of FIELD_MODULE)
        to be used for a tag with obj as value, if a datatype has not
        been specified by the user
    """
    if getattr(obj, "_default_gfa_tag_datatype",None):
      return obj._default_gfa_tag_datatype()
    else:
      if isinstance(obj, list) and\
             (all([isinstance(v, builtins.int) for v in obj]) or
              all([isinstance(v, builtins.float) for v in obj])):
        return "B"
      for k,v in gfapy.Field._default_tag_datatypes:
        if isinstance(obj, k):
          return v
      return "J"

  @classmethod
  def register_datatype(cls, name, klass):
    """Register a custom-defined datatype class

    Parameters:
      name (str) : the identifier to be used for the datatype. This is
        to be used in the field datatype declaration of extensions
        definining custom records, which use this custom datatype
      klass (class) : the class which provide the decode, encode,
        validate_encoded and validate_decoded methods for handling
        data of the custom datatype
    """
    cls.GFA2_POSFIELD_DATATYPE.append(name)
    cls.FIELD_MODULE[name] = klass

